paywall/src/__tests__/data-iframe/start/makePurchaseCallback.test.js

```diff
@@ -0,0 +1,128 @@
+import makePurchaseCallback from '../../../data-iframe/start/makePurchaseCallback'
+import {
+  setLocks,
+  setKeys,
+  setAccount,
+  setNetwork,
+} from '../../../data-iframe/cacheHandler'
+import { setAccount as setBlockchainAccount } from '../../../data-iframe/blockchainHandler/account'
+
+describe('makePurchaseCallback', () => {
+  let walletService
+  let web3Service
+  const requiredConfirmations = 2
+  let update
+  let fakeWindow
+
+  it('should return a function that can be used to purchase a key', () => {
+    expect.assertions(1)
+
+    const purchase = makePurchaseCallback({
+      walletService,
+      web3Service,
+      requiredConfirmations,
+      update,
+      window: fakeWindow,
+    })
+
+    expect(purchase).toBeInstanceOf(Function)
+  })
+
+  describe('purchase callback', () => {
+    let purchase
+
+    beforeEach(async () => {
+      fakeWindow = {
+        storage: {},
+        localStorage: {
+          setItem(key, item) {
+            fakeWindow.storage[key] = item
+          },
+          getItem(key) {
+            return fakeWindow.storage[key]
+          },
+          removeItem(key) {
+            delete fakeWindow.storage[key]
+          },
+        },
+      }
+
+      web3Service = {}
+      walletService = {
+        ready: true,
+        addListener: jest.fn(),
+        removeListener: jest.fn(),
+        once: jest.fn(),
+        on: jest.fn(),
+        purchaseKey: jest.fn().mockResolvedValue(),
+      }
+      update = jest.fn()
+
+      purchase = makePurchaseCallback({
+        walletService,
+        web3Service,
+        requiredConfirmations,
+        update,
+        window: fakeWindow,
+      })
+
+      await setAccount(fakeWindow, 'account')
+      setBlockchainAccount('account') // this is handled in the "ensureWalletReady()" portion
+      await setNetwork(fakeWindow, 2)
+      await setLocks(fakeWindow, {
+        lock: {
+          address: 'lock',
+          keyPrice: '123',
+        },
+      })
+      await setKeys(fakeWindow, {
+        lock: {
+          lock: 'lock',
+          owner: 'account',
+        },
+      })
+    })
+
+    it('should initiate a key purchase', async () => {
+      expect.assertions(1)
+
+      await purchase('lock')
+
+      await new Promise(resolve => {
+        const interval = setInterval(() => {
+          if (walletService.purchaseKey.mock.calls.length) {
+            clearInterval(interval)
+            resolve()
+          }
+        })
+      })
+      expect(walletService.purchaseKey).toHaveBeenCalledWith(
+        'lock',
+        'account',
+        '123'
+      )
+    })
+
+    it('should initiate monitoring of key purchase transaction', async () => {
```

Comment:
Would it be the right place to test what happens when the key purchase transaction actually fails to be sent?

Reply:
yes, we can test that throw errors cascade out